### PR TITLE
Add interpreter for DotGeneralOp

### LINKED ISSUE #336: Add interpreter for dot_general

### PR DESCRIPTION
**Pull request by ghpvnist**
We have the following non-quantization-related constraints (excluding C13, C15-C20) in the spec:

```
(I1) lhs tensor.
(I2) rhs tensor.
(I3) lhs_batching_dimensions 1-dimensional tensor constant of type `si64`.
(I4) rhs_batching_dimensions 1-dimensional tensor constant of type `si64`.
(I5) lhs_contracting_dimensions 1-dimensional tensor constant of type `si64`.
(I6) rhs_contracting_dimensions 1-dimensional tensor constant of type `si64`.
(I7) precision_config variadic number of enum of `DEFAULT`, `HIGH`, and `HIGHEST`.
(C1) size(`lhs_batching_dimensions`) = size(`rhs_batching_dimensions`).
(C2) size(`lhs_contracting_dimensions`) =
size(`rhs_contracting_dimensions`).
(C3) `lhs_batching_dimensions` and `lhs_contracting_dimensions` combined are
unique.
(C4) `rhs_batching_dimensions` and `rhs_contracting_dimensions` combined are
unique.
(C5) 0 <= `lhs_batching_dimensions[i]` < rank(`lhs`) for all `i`
in [0, size(`lhs_batching_dimensions`)).
(C6) 0 <= `lhs_contracting_dimensions[i]` < rank(`lhs`) for all `i`
in [0, size(`lhs_contracting_dimensions`)).
(C7) 0 <= `rhs_batching_dimensions[i]` < rank(`rhs`) for all `i`
in [0, size(`rhs_batching_dimensions`)).
(C8) 0 <= `rhs_contracting_dimensions[i]` < rank(`rhs`) for all `i`
in [0, size(`rhs_contracting_dimensions`)).
(C9) dim(`lhs`, `lhs_batching_dimensions[i]`) =
dim(`rhs`, `rhs_batching_dimensions[i]`) for all `i` in [0,
size(`lhs_batching_dimensions`)).
(C10) dim(`lhs`, `lhs_contracting_dimensions[i]`) =
dim(`rhs`, `rhs_contracting_dimensions[i]`) for all `i` in [0,
size(`lhs_contracting_dimensions`)).
(C11) size(`precision_config`) = 2.
(C12) shape(`result`) = dim(`lhs`, `lhs_batching_dimensions`) +
dim(`lhs`, `lhs_result_dimensions`) + dim(`rhs`, `rhs_result_dimensions`).
(C14) element_type(`lhs`) = element_type(`rhs`).
```

These constraints will be comprehensively covered by the following tests:

```
I1: a) lhs is not a tensor. (Covered by ODS).
I2: a) rhs is not a tensor. (Covered by ODS).
I3: a) lhs_batching_dimensions is not a 1-dimensional tensor. (Covered by ODS).
    b) element_type(lhs_batching_dimesnions) != `si64`. (Covered by ODS).
I4: a) rhs_batching_dimensions is not a 1-dimensional tensor. (Covered by ODS).
    b) element_type(rhs_batching_dimesnions) != `si64`. (Covered by ODS).
I5: a) lhs_contracting_dimensions is not a 1-dimensional tensor. (Covered by ODS).
    b) element_type(lhs_contracting_dimensions) != `si64`. (Covered by ODS).
I6: a) rhs_contracting_dimensions is not a 1-dimensional tensor. (Covered by ODS).
    b) element_type(rhs_contracting_dimensions) != `si64`. (Covered by ODS).
I7: a) precision_config does not have variadic number of enum of `DEFAULT`, `HIGH`, and `HIGHEST`. (Covered by ODS).
C1: a) size(lhs_batching_dimensions) != size(rhs_batching_dimensions).
C2: a) size(lhs_contracting_dimensions) != size(rhs_contracting_dimensions).
C3: a) lhs_batching_dimensions and lhs_contracting_dimensions combined are not unique.
C4: a) rhs_batching_dimensions and rhs_contracting_dimensions combined are not unique.
C5: a) lhs_batching_dimensions[i] < 0 for any i.
    b) lhs_batching_dimensions[i] >= rank(lhs) for any i.
C6: a) lhs_contracting_dimensions[i] < 0 for any i.
    b) lhs_contracting_dimensions[i] >= rank(lhs) for any i.
C7: a) rhs_batching_dimensions[i] < 0 for any i.
    b) rhs_batching_dimensions[i] >= rank(rhs) for any i.
C8: a) rhs_contracting_dimensions[i] < 0 for any i.
    b) rhs_contracting_dimensions[i] >= rank(rhs) for any i.
C9: a) dim(lhs, lhs_batching_dimensions[i]) != dim(rhs, rhs_batching_dimensions[i]) for any i.
C10: a) dim(lhs, lhs_contracting_dimensions[i]) != dim(rhs, rhs_contracting_dimensions[i]) for any i.
C11: a) size(precision_config) != 2.
C12: no negative test needed since it's just inferring the shape.
C14: a) element_type(lhs) != element_type(rhs).
```

If we drop the "Covered by ODS" pieces, this will leave us with the following test cases:

```
C1a: size(lhs_batching_dimensions) != size(rhs_batching_dimensions).
C2a: size(lhs_contracting_dimensions) != size(rhs_contracting_dimensions).
C3a: lhs_batching_dimensions and lhs_contracting_dimensions combined are not unique.
C4a: rhs_batching_dimensions and rhs_contracting_dimensions combined are not unique.
C5a: lhs_batching_dimensions[i] < 0 for any i.
C5b: lhs_batching_dimensions[i] >= rank(lhs) for any i.
C6a: lhs_contracting_dimensions[i] < 0 for any i.
C6b: lhs_contracting_dimensions[i] >= rank(lhs) for any i.
C7a: rhs_batching_dimensions[i] < 0 for any i.
C7b: rhs_batching_dimensions[i] >= rank(rhs) for any i.
C8a: rhs_contracting_dimensions[i] < 0 for any i.
C8b: rhs_contracting_dimensions[i] >= rank(rhs) for any i.
C9a: dim(lhs, lhs_batching_dimensions[i]) != dim(rhs, rhs_batching_dimensions[i]) for any i.
C10a: dim(lhs, lhs_contracting_dimensions[i]) != dim(rhs, rhs_contracting_dimensions[i]) for any i.
C11a: size(precision_config) != 2.
C14a: element_type(lhs) != element_type(rhs).
```

Notes:
* (C14) currently does not have a test and consider removing it #755
* (C11) currently does not have a test due to #755 and #879.

closes #336

### CODE REVIEW

**Comment by ghpvnist** [comment]
C12 currently does not have a test due to #755 and #879

File: stablehlo/dialect/StablehloOps.td [hunk 1]
```diff
@@ -2200,7 +2200,7 @@ def StableHLO_DotOp: StableHLO_Op<"dot", [Pure]> {
 }
 
 def StableHLO_DotGeneralOp: StableHLO_ShapedInterfaceOp<"dot_general", [Pure]> {
-  let summary = "General Dot operator";
+  let summary = "DotGeneral operation";
```
Comment (sdasgup3) [inline stablehlo/dialect/StablehloOps.td:2203]: Add input constraints as per  the checklist.

**Comment by ghpvnist** [comment]
This is not a reference implementation to the spec, but decided to keep it as the reference implementation requires reliance on MLIR builders (to create ReduceOp) which affects readability. Thoughts? @burmako

**Comment by burmako** [comment]
Happy to review as is, and then let's discuss that as part of the review?

File: stablehlo/tests/interpret_dot_general.mlir [hunk 2]
```diff
@@ -0,0 +1,103 @@
+// RUN: stablehlo-translate --interpret -split-input-file %s
```
Comment (burmako) [inline stablehlo/tests/interpret_dot_general.mlir:1]: Also, we need tests for all different control flow paths through the implementation. Just one test is not going to be sufficient.

File: stablehlo/reference/Ops.cpp [hunk 3]
```diff
@@ -527,6 +543,73 @@ Tensor evalDivideOp(const Tensor &lhs, const Tensor &rhs,
...
+        if (lhsIdx[lhsContractingDimensions[j]] != contractingDimSizes[j])
+          break;
+        lhsIdx[lhsContractingDimensions[j]] = 0;
+        rhsIdx[rhsContractingDimensions[j]] = 0;
+      }
+    }
+    result.set(*resultIt, sum);
+  }
+  return result;
+}
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:611]: I support diverging this implementation from the spec, which is way too complicated currently. Opened #1551 to track bringing the two into alignment, but I don't think this is a high priority.

Comment (burmako) [inline stablehlo/reference/Ops.cpp:611]: Proposed several refactorings to this code in https://github.com/ghpvnist/stablehlo/pull/6.

**Review by burmako — CHANGES_REQUESTED** [review]

**Comment by burmako** [comment]
@sdasgup3 I've only reviewed Ops.cpp. Can you review the rest please?

File: stablehlo/reference/Ops.cpp [hunk 4]
```diff
@@ -527,6 +543,73 @@ Tensor evalDivideOp(const Tensor &lhs, const Tensor &rhs,
...
+
+    // The non-batching, non-contracting dimensions of the operands are copied
+    // over to the result dimensions following the batching dimensions.
+    for (uint64_t i = 0; i < lhsNonBatchingNonContractingDims.size(); i++)
+      lhsIdx[lhsNonBatchingNonContractingDims[i]] = (*resultIt)[resultDim++];
+
+    for (uint64_t i = 0; i < rhsNonBatchingNonContractingDims.size(); i++)
+      rhsIdx[rhsNonBatchingNonContractingDims[i]] = (*resultIt)[resultDim++];
+
+    auto sum = Element(resultType.getElementType(), 0.0);
```
Comment (burmako) [inline stablehlo/reference/Ops.cpp:594]: This reminds me that in the spec, instead of `0`, we should be using `constant(0, element_type(result))`.

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/reference/Ops.cpp (hunk 3 above)
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:1070]: Thanks for the suggestions!

File: stablehlo/dialect/TypeInference.cpp [hunk 5]
```diff
@@ -1991,7 +1999,7 @@ LogicalResult inferDotGeneralOp(
     if (!llvm::is_contained(rhsBatchingDimensions, i) &&
         !llvm::is_contained(rhsContractingDimensions, i))
       dimensions.push_back(rhsShape[i]);
-
+  // dot_general_c12
   inferredReturnShapes.emplace_back(dimensions);
   return success();
 }
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:2005]: The PR description says "(C14) currently does not have a test and consider removing it Allow distinct precision_config values in DotGeneralOp #1354", but it doesn't look like #1354 is related to C14?

File: stablehlo/dialect/StablehloOps.td [hunk 6]
```diff
@@ -2220,15 +2220,17 @@ def StableHLO_DotGeneralOp: StableHLO_ShapedInterfaceOp<"dot_general", [Pure]> {
...
     ```
   }];
   let arguments = (ins
-    HLO_Tensor:$lhs,
-    HLO_Tensor:$rhs,
-    StableHLO_DotDimensionNumbers:$dot_dimension_numbers,
-    StableHLO_PrecisionConfigAttr:$precision_config
+    HLO_Tensor:$lhs /*dot_general_i1*/,
+    HLO_Tensor:$rhs /*dot_general_i2*/,
+    StableHLO_DotDimensionNumbers:$dot_dimension_numbers /*dot_general_i3, dot_general_i4, dot_general_i5, dot_general_i6*/,
```
Comment (burmako) [inline stablehlo/dialect/StablehloOps.td:2232]: The PR description says that i3a, i4a, i5a and i6a are not covered by the ODS, but I think they are.

File: stablehlo/dialect/StablehloOps.td [hunk 7]
```diff
@@ -2220,15 +2220,17 @@ def StableHLO_DotGeneralOp: StableHLO_ShapedInterfaceOp<"dot_general", [Pure]> {
...
+    StableHLO_PrecisionConfigAttr:$precision_config /*dot_general_i7*/
```
Comment (burmako) [inline stablehlo/dialect/StablehloOps.td:2233]: Ditto for i7a.

File: stablehlo/tests/infer_stablehlo.mlir [hunk 8]
```diff
@@ -1508,9 +1508,9 @@ func.func @real_dynamic_slice(%arg0: tensor<?xf32>, %arg1: tensor<1xindex>, %arg
 
 // -----
 
-// CHECK-LABEL: func @dot_general
+// CHECK-LABEL: func @dot_general_c12
```
Comment (burmako) [inline stablehlo/tests/infer_stablehlo.mlir:1511]: Let's keep `dot_general` given that I don't think we'll be distinguishing test cases in this file by constraint names.

File: stablehlo/tests/interpret_dot_general.mlir [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+func.func @dot_general_op_test_si64() {
```
Comment (burmako) [inline stablehlo/tests/interpret_dot_general.mlir:3]: Can we simplify these names from `dot_general_op_test_foo` to just `dot_general_foo`? Or even better, just `foo`? Not sure what the current convention is.

File: stablehlo/tests/ops_stablehlo.mlir [hunk 10]
```diff
@@ -2983,12 +2986,12 @@ func.func @dot_general(%arg0: tensor<?x?x?xf32>, %arg1: tensor<?x?x?xf32>) {
...
     >
   } : (tensor<?x?x?xf32>, tensor<?x?x?xf32>) -> tensor<?x?x?xf32>
-  func.return
+  func.return %0 : tensor<?x?x?xf32>
 }
 
 // -----
 
-func.func @dot_general(%arg0: tensor<*xf32>, %arg1: tensor<*xf32>) {
+func.func @dot_general_c1(%arg0: tensor<*xf32>, %arg1: tensor<*xf32>) -> tensor<*xf32>{
```
Comment (burmako) [inline stablehlo/tests/ops_stablehlo.mlir:2994]: Do we need two tests for C1? (Same for other constraints below, e.g. C2 and C3).

File: stablehlo/tests/ops_stablehlo.mlir [hunk 11]
```diff
@@ -3298,59 +3301,49 @@ func.func @dot_general(%arg0: tensor<?x2x?xf32>, %arg1: tensor<?x3x?xf32>) {
...
-    precision_config = [#stablehlo<precision DEFAULT>, #stablehlo<precision DEFAULT>, #stablehlo<precision DEFAULT>]
+    precision_config = [#stablehlo<precision DEFAULT>]
   } : (tensor<2x3x4xf32>, tensor<2x3x5xf32>) -> tensor<2x4x5xf32>
   func.return %0 : tensor<2x4x5xf32>
 }
 
 // -----
 
-func.func @dynamic_reshape(%arg0: tensor<?xf32>, %shape: tensor<2xindex>) -> tensor<?x?xf32> {
+func.func @compatible_shapes(%arg0: tensor<?xf32>, %shape: tensor<2xindex>) -> tensor<?x?xf32> {
```
Comment (burmako) [inline stablehlo/tests/ops_stablehlo.mlir:3346]: Unrelated change?

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/tests/interpret_dot_general.mlir (hunk 9 above)
Comment (ghpvnist) [reply stablehlo/tests/interpret_dot_general.mlir:3]: Current convention is to have `dot_general_op_test_foo`. I agree this is a long and unnecessary name. Maybe it's better to rename it such that (in order) the test copied from spec should be `spec_example` or `example`, ones testing element types be `si64` or something equivalent, and the rest be some concise name that doesn't include the obvious prefixes. I'll create a ticket for that and keep the name for now as this is the current convention.

File: stablehlo/tests/ops_stablehlo.mlir (hunk 11 above)
Comment (ghpvnist) [reply stablehlo/tests/ops_stablehlo.mlir:3346]: Thanks!

File: stablehlo/tests/ops_stablehlo.mlir (hunk 10 above)
Comment (ghpvnist) [reply stablehlo/tests/ops_stablehlo.mlir:2994]: No, these should be removed according to the latest checklist.

File: stablehlo/tests/interpret_dot_general.mlir (hunk 9 above)
Comment (ghpvnist) [reply stablehlo/tests/interpret_dot_general.mlir:3]: Issue: #1596

File: stablehlo/tests/infer_stablehlo.mlir (hunk 8 above)
Comment (ghpvnist) [reply stablehlo/tests/infer_stablehlo.mlir:1511]: The current convention in the [checklist 6.i](https://github.com/openxla/stablehlo/blob/main/docs/reference_checklist.md) is to also tag the tests with constraints in this file. There are other ops whose constraints do matter (mainly batch_norm_* ops). If we decide to remove it, it's probably best to address them in bulk.

File: stablehlo/dialect/StablehloOps.td (hunk 6 above)
Comment (ghpvnist) [reply stablehlo/dialect/StablehloOps.td:2232]: Ack.

File: stablehlo/dialect/StablehloOps.td (hunk 7 above)
Comment (ghpvnist) [reply stablehlo/dialect/StablehloOps.td:2233]: Ack.

File: stablehlo/dialect/TypeInference.cpp (hunk 5 above)
Comment (ghpvnist) [reply stablehlo/dialect/TypeInference.cpp:2005]: This ticket originally tracked it, but was renamed/merged with a different ticket. Updated the link to point to the right one.

File: stablehlo/dialect/TypeInference.cpp [hunk 12]
```diff
@@ -1908,14 +1911,14 @@ LogicalResult inferDotGeneralOp(
     }
     return success();
   };
-
+  // dot_general_c3
```
Comment (burmako) [inline stablehlo/dialect/TypeInference.cpp:1914]: Let's follow the Google C++ Style Guide regarding vertical whitespace: https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace?

"A blank line before a comment line usually helps readability — the introduction of a new comment suggests the start of a new thought, and the blank line makes it clear that the comment goes with the following thing instead of the preceding".

File: stablehlo/dialect/StablehloOps.td (hunk 6 above)
Comment (burmako) [reply stablehlo/dialect/StablehloOps.td:2232]: These test cases still show up under "If we drop the "Covered by ODS" pieces, this will leave us with the following test cases".

File: stablehlo/tests/ops_stablehlo.mlir [hunk 13]
```diff
@@ -3073,27 +3031,12 @@ func.func @dot_general(%arg0: tensor<?x?x?xf32>, %arg1: tensor<?x?x?xf32>) {
...
-    >
-  } : (tensor<*xf32>, tensor<*xf32>) -> tensor<*xf32>
-  func.return
+  func.return %0 : tensor<?x?x?xf32>
 }
 
 // -----
 
-func.func @dot_general(%arg0: tensor<?x?x?xf32>, %arg1: tensor<?x?x?xf32>) {
+func.func @dot_general_c3(%arg0: tensor<?x?x?xf32>, %arg1: tensor<?x?x?xf32>) -> tensor<?x?x?xf32> {
```
Comment (burmako) [inline stablehlo/tests/ops_stablehlo.mlir:3039]: Duplicate test for dot_general_c3?

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/dialect/StablehloOps.td (hunk 6 above)
Comment (ghpvnist) [reply stablehlo/dialect/StablehloOps.td:2232]: They should be gone and may not be updated on your side. Try refreshing the page?

File: stablehlo/dialect/TypeInference.cpp (hunk 12 above)
Comment (ghpvnist) [reply stablehlo/dialect/TypeInference.cpp:1916]: Done.

File: stablehlo/tests/ops_stablehlo.mlir (hunk 13 above)
Comment (ghpvnist) [reply stablehlo/tests/ops_stablehlo.mlir]: Done.

**Review by burmako — APPROVED** [review]
